### PR TITLE
Implemented realization of RegistrationServiceImpl and RegistrationSe…

### PR DESCRIPTION
**Pull request by Trayvor**
…rviceImplTest with tests for that class

### CODE REVIEW

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 1]
```diff
@@ -1,14 +1,43 @@
...
+    public static final int MIN_PASSWORD_LENGTH = 6;
     private final StorageDao storageDao = new StorageDaoImpl();
 
     @Override
-    public User register(User user) {
-        return null;
+    public User register(User user) throws InvalidDataException {
+        if (storageDao.get(user.getLogin()) != null) {
+            return null;
+        }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:18]: Bad practice. Just treat as other checks - throw your exception with an informative message. Place it after user.getLogin null check

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 2]
```diff
@@ -1,14 +1,43 @@
...
+        if (user.getLogin() == null) {
+            throw new InvalidDataException("Login is null");
+        }
+        if (user.getLogin().length() < MIN_LOGIN_LENGTH) {
+            throw new InvalidDataException("Login is too small,"
+                    + " minimal login length is " + MIN_LOGIN_LENGTH + " characters");
+        }
+        if (user.getAge() == null) {
+            throw new InvalidDataException("Age is incorrect");
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:35]:
```suggestion
            throw new InvalidDataException("Age is null");
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        user1.setAge(-10);
+        Assertions.assertThrows(InvalidDataException.class,
+                () -> registrationService.register(user1));
+        user1.setAge(-1);
+        Assertions.assertThrows(InvalidDataException.class,
+                () -> registrationService.register(user1));
+    }
+
+    @Test
+    void register_insufficientAge_notOk() {
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:69]:
```suggestion
    void register_underage_notOk() {
```
we call ppl who haven't reached 18 underage

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+    @Test
+    void register_sufficientAge_Ok() throws InvalidDataException {
+        Assertions.assertEquals(user1, registrationService.register(user1));
+        user2.setAge(19);
+        Assertions.assertEquals(user2, registrationService.register(user2));
+        user3.setAge(100);
+        Assertions.assertEquals(user3, registrationService.register(user3));
+        user4.setAge(Integer.MAX_VALUE);
+        Assertions.assertEquals(user4, registrationService.register(user4));
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:89]: we need only one _ok test - where all fields are valid. You can add a second test to test multiple consecutive registrations, but it's optional

File: src/main/java/core/basesyntax/InvalidDataException.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package core.basesyntax;
+
+public class InvalidDataException extends Exception {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/InvalidDataException.java:3]: image

File: src/main/java/core/basesyntax/service/RegistrationService.java [hunk 6]
```diff
@@ -1,7 +1,8 @@
 package core.basesyntax.service;
 
+import core.basesyntax.InvalidDataException;
 import core.basesyntax.model.User;
 
 public interface RegistrationService {
-    User register(User user);
+    User register(User user) throws InvalidDataException;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/RegistrationService.java:7]:
```suggestion
    User register(User user);
```
fix everywhere

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
+    private static RegistrationServiceImpl registrationService;
+    private User user1;
+    private User user2;
+    private User user3;
+    private User user4;
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:16]: naming with numbers is a bad practice. You need only one field - a valid user. Break validation in each test separately having a valid user as a base - set password too short etc.

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        Assertions.assertEquals(user2, registrationService.register(user2));
+        user3.setPassword("1234567891011121314151617");
+        Assertions.assertEquals(user3, registrationService.register(user3));
+    }
+
+    @Test
+    void register_addingExistingUser_notOk() throws InvalidDataException {
+        registrationService.register(user1);
+        Assertions.assertNull(registrationService.register(user1));
+    }
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:166]: you can't rely on register to test register
image

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+    @Test
+    void register_addingSameUser_notOk() throws InvalidDataException {
+        registrationService.register(user1);
+        user2.setLogin(user1.getLogin());
+        Assertions.assertNull(registrationService.register(user2));
+    }
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:173]: we covered it already

**Review by okuzan — CHANGES_REQUESTED** [review]
Good job overall!

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        defaultUser.setAge(18);
+        defaultUser.setLogin("username1");
+        defaultUser.setPassword("123456");
+    }
+
+    @Test
+    void register_nullAge_notOk() {
+        defaultUser.setAge(null);
+        Assertions.assertThrows(InvalidDataException.class,
+                () -> registrationService.register(defaultUser));
```
Comment (olekskov) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:34]: I'd recommend to add static import for heavily used static method in order not to indicate each time class used
```suggestion
        assertThrows(InvalidDataException.class,
                () -> registrationService.register(defaultUser));
```
and
`import static org.junit.jupiter.api.Assertions.assertThrows;`

**Review by olekskov — APPROVED** [review]
Good job, approved, please check comment for further

**Pull request closed without merging**